### PR TITLE
IGNITE 6699-2

### PR DESCRIPTION
**Pull request by 1vanan**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 1]
```diff
@@ -653,6 +690,92 @@ else if (autoFlushFreq == 0)
...
+    /**
+     * @param entry Entry.
+     * @return Future.
+     */
+    private IgniteFuture<?> addDataInternal(DataStreamerEntry entry) {
+        enterBusy();
+
+        streamerLock.readLock().lock();
+
+        Collection<DataStreamerEntry> entries0;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: All code after `lock` has to be under `try {} finally {lock.unlock();}` statement.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 2]
```diff
@@ -653,6 +690,92 @@ else if (autoFlushFreq == 0)
...
+        }
+
+        fut0 = threadBufMap.get(Thread.currentThread().getId()).get1();
+
+        try {
+            threadBufMap.get(Thread.currentThread().getId()).get2().add(entry);
+
+            entries0 = threadBufMap.get(Thread.currentThread().getId()).get2();
+
+            if (!activeFuts.contains(fut0.internalFuture())) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: Seems like this check can be moved to future creation block.
Why we has to check `activeFuts.contains` every time?
We can add `fut0` to `activeFuts` on creation.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 3]
```diff
@@ -1041,6 +1167,19 @@ private void doFlush() throws IgniteCheckedException {
 
         int doneCnt = 0;
 
+        streamerLock.writeLock().lock();
+
+        for (Long threadId :
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: All operations after `lock()` has to be under `try {} finally {lock.unlock();}`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 4]
```diff
@@ -1168,6 +1307,18 @@ else if (f.isDone()) {
             return;
 
         try {
+            streamerLock.writeLock().lock();
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: All operations after lock() has to be under try {} finally {lock.unlock();}

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 5]
```diff
@@ -662,6 +785,9 @@ else if (autoFlushFreq == 0)
 
     /** {@inheritDoc} */
     @Override public IgniteFuture<?> addData(K key, V val) {
+        if (closedStreamer)
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: This check has to be added to all Streamer methods.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 6]
```diff
@@ -653,6 +690,92 @@ else if (autoFlushFreq == 0)
...
+    private IgniteFuture<?> addDataInternal(DataStreamerEntry entry) {
+        enterBusy();
+
+        streamerLock.readLock().lock();
+
+        Collection<DataStreamerEntry> entries0;
+
+        IgniteCacheFutureImpl fut0;
+
+        if (!threadBufMap.containsKey(Thread.currentThread().getId())) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: Please, create variable `long threadId = Thread.currentThread().getId()` and use it through the code.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 7]
```diff
@@ -653,6 +690,92 @@ else if (autoFlushFreq == 0)
...
+     * @return Future.
+     */
+    private IgniteFuture<?> addDataInternal(DataStreamerEntry entry) {
+        enterBusy();
+
+        streamerLock.readLock().lock();
+
+        Collection<DataStreamerEntry> entries0;
+
+        IgniteCacheFutureImpl fut0;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: Please, declare a variable in the place where it used.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java (hunk 1 above)
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: Please, declare a variable in the place where it used.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 8]
```diff
@@ -653,6 +690,92 @@ else if (autoFlushFreq == 0)
...
     }
 
+    /**
+     * @param entry Entry.
+     * @return Future.
+     */
+    private IgniteFuture<?> addDataInternal(DataStreamerEntry entry) {
+        enterBusy();
+
+        streamerLock.readLock().lock();
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: Please, cache `Lock` variable
`Lock readLock = streamerLock.readLock();`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 9]
```diff
@@ -653,6 +690,92 @@ else if (autoFlushFreq == 0)
...
+            IgniteBiTuple<IgniteCacheFutureImpl, List<DataStreamerEntry>> futBufPair =
+                    new IgniteBiTuple(futPerBatch, entriesPerBatch);
+
+            threadBufMap.put(Thread.currentThread().getId(), futBufPair);
+        }
+
+        fut0 = threadBufMap.get(Thread.currentThread().getId()).get1();
+
+        try {
+            threadBufMap.get(Thread.currentThread().getId()).get2().add(entry);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: Please, cache results of `threadBufMap.get(Thread.currentThread().getId()).get2()` into some variable

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java (hunk 8 above)
Comment (1vanan) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: Do you mean that everywhere than lock is used, I should make something like this: 
        Lock readLock = streamerLock.readLock();
        readLock.lock();
        try{
//some action
        }
        finally{
            readLock.unlock();
        }

Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: Yes.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java (hunk 5 above)
Comment (1vanan) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: like doFlush(), tryFlush() and close()?

Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: to all operations that can't invoke on closed stream.
Is there such check for now?
Can we call some method on a closed streamer?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java (hunk 8 above)
Comment (1vanan) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: And for writeLock strategy is similar?

Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: Yes

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java (hunk 2 above)
Comment (1vanan) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: There are two future creation blocks: in `addDataInternal` and in `clearBuf`. When we invoke `addData`, we don't know there are other entries in buffer or buffer is empty. And if we add empty future to `activeFuts`, failure will occur: at least one empty future will be in `activeFuts` at the end.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 10]
```diff
@@ -653,6 +686,87 @@ else if (autoFlushFreq == 0)
...
+                IgniteCacheFutureImpl futPerBatch = new IgniteCacheFutureImpl(new GridFutureAdapter());
+
+                List<DataStreamerEntry> entriesPerBatch = new ArrayList<>(bufLdrSizePerBatch);
+
+                IgniteBiTuple<IgniteCacheFutureImpl, List<DataStreamerEntry>> futBufPair =
+                        new IgniteBiTuple(futPerBatch, entriesPerBatch);
+
+                activeFuts.add(futPerBatch.internalFuture());
+
+                futPerBatch.internalFuture().listen(rmvActiveFut);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: line 713 and 711 should be reordered.

First, you fully initialize Future.
Second, you add it in `activeFuts`.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 11]
```diff
@@ -1041,6 +1158,23 @@ private void doFlush() throws IgniteCheckedException {
 
         int doneCnt = 0;
 
+        writeLock.lock();
+
+        try {
+            for (Long threadId :
+                    threadBufMap.keySet()) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: Please, write a`for (Long threadId : threadBufMap.keySet()) {` on a single line.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 12]
```diff
@@ -1168,6 +1302,22 @@ else if (f.isDone()) {
             return;
 
         try {
+            writeLock.lock();
+
+            try {
+                for (Long threadId :
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: Please, write `for (Long threadId : threadBufMap.keySet()) {` on a single line.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java [hunk 13]
```diff
@@ -1168,6 +1302,22 @@ else if (f.isDone()) {
...
+                        threadBufMap.keySet()) {
+                    List <DataStreamerEntry> entries0 = threadBufMap.get(threadId).get2();
+
+                    if (entries0.size() > 0)
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImpl.java]: As far as I can understand we can emit this `if` - entries size is always greater then 0.

**Pull request closed without merging**